scripts/ui.ts

```diff
@@ -106,57 +109,154 @@ async function init(): Promise<void> {
+  const permissionsResponse = await getPermissions(seed, permissions);
+
+  // Return remaining failed permissions to skapp.
+  log("Returning permissions response");
+  return [true, permissionsResponse];
+}
+
+// ==========
+// Core Logic
+// ==========
+
+/**
+ * Check if the browser is supported and throw an error if not.
+ *
+ * @throws - Will throw if the browser is not supported.
+ */
+async function checkBrowserSupported(): Promise<void> {
   const [isSupported, reason] = await MySky.isBrowserSupported();
   if (!isSupported) {
     alert(reason);
     throw new Error(reason);
   }
+}
 
-  // Get the seed and email.
-
+/**
+ * Gets the seed and email. The flow is:
+ *
+ * 1. Check for the seed in local storage.
+ *
+ * 2. If the seed is not found, get it and maybe the email from a seed provider.
+ *
+ * 3. If we did not open the provider, or it did not provide the email, then
+ * check for an email in the saved user settings.
+ *
+ * 4. If we still didn't get an email, and the seed provider action was signin,
+ * then we display the signin-connect page where the user may connect his email
+ * on signin.
+ *
+ * 5. If we got the email, then we register/login to set the JWT cookie.
+ *
+ * 6. If the user provided a new email at some point, then we save it in user
+ * settings, after having successfully connected to a portal account.
+ *
+ * (7. We return the seed and email and save them in storage in another
+ * function, which triggers Main MySky's storage listener.)
+ *
+ * @returns - The seed and email.
+ */
+async function getSeedAndEmail(): Promise<[Uint8Array, string | null]> {
   let seed = checkStoredSeed();
+  let email: string | null = null;
+  let action: SeedProviderAction | null = null;
+  let emailProvidedByUser = null;
 
-  // If we don't have a seed, get it from the seed provider. We will also get
+  // If we don't have a seed, get it from the seed provider. We may also get
   // the email if the user is signing up for the first time.
   if (!seed) {
-    // Show seed provider chooser.
-
-    const seedProviderDisplayUrl = await getSeedProviderDisplayUrl();
-
-    // User has chosen seed provider, open seed provider display.
-
-    log("Calling runSeedProviderDisplay");
-    let email = null;
-    [seed, email] = await runSeedProviderDisplay(seedProviderDisplayUrl);
-
-    // Register and get the JWT.
-    let jwt = null;
-    if (email) {
-      try {
-        jwt = await register(client, seed, email);
-      } catch (e1) {
-        try {
-          jwt = await login(client, seed, email);
-        } catch (e2) {
-          throw new Error(`Could not register: ${e1}. Could not login: ${e2}`);
-        }
-      }
+    const resp = await getSeedAndEmailFromProvider();
+    [seed, email, action] = [resp.seed, resp.email, resp.action];
+    emailProvidedByUser = email !== null;
+  }
+
+  if (action === "signin") {
+    // Assert that we did not get an email from the signin page.
+    if (emailProvidedByUser) {
```

Comment:
I'll add a comment. If you still think this is weird I can maybe rename the function, but nothing clearly better jumps to mind.

Reply:
OK, so, these parameters don't seem to be coming from any storage (localStorage, session, cookie, etc.) but instead they come from the request coming from the skapp that runs in a different window/iframe. Is that correct? This is the part that I was missing here. Because it seemed to me that we're fetching those from our own storage where it seems logical for us to already have the email. But if the "data source" is the request made by the skapp then this makes more sense.

Reply:
Yes, we only request the email in case we don't get the seed from local storage and we have to display the seed provider. If we do get the seed from storage, then we get the stored email from user settings stored on MySky instead of relying on local storage.